### ISSUE TITLE
cdylib still not fully functional

### ISSUE THREAD

**Issue by MarkSwanson** [rust-lang/cargo#2868]
cdylib works perfectly if I do this in my build.rs:

    println!("cargo:rustc-link-search=.../target/debug/deps/");
    println!("cargo:rustc-link-lib=my_custom_ffi-b248e1a221d9ae20");

The problem is that I can not hard code the hash. If I don't hard code the hash cargo will fail.

Is there a public API available to calculate this hash?
I can use cargo pkgid my_custom_ffi . Perhaps there is a standard hash Fn I can use to get the hash?

There seems to be no other way to create a cdylib that links dynamically against a dependent DSO. ?

Thanks!

**Comment by alexcrichton** [comment]
Ah yeah the intention of cdylib was that other Rust crates wouldn't depend on it, it's the endpoint of a compilation like an executable or a staticlib artifact.

Could you describe how come you're linking a Rust cdylib into another Rust library? It sounds like the Rust cdylib may want to be a totally independent Cargo project that's hooked up externally rather than through Cargo itself.

**Comment by MarkSwanson** [comment]
1. create a Rust shared library that contains common functionality.
2. create n Rust shared libraries/plugins that run in separate containers - there will be a good number of these and because of memory constraints we need these plugins to dynamically link against the common shared library.

I really don't mind calculating the hash myself.
cargo pkgid project gives me the pkgid which I think is hashed. Maybe cargo exposes the hash Fn and I just call it?

**Comment by MarkSwanson** [comment]
I found a way that seems to work. In my plugin, create a make.sh file that simply:

cargo rustc --lib -- -C prefer-dynamic

Bonus: I can specify in my Cargo.toml:
crate-type = ["dylib", "rlib"]

This is important because without rlib none of the tests will build.

I _thought_ that cdylib was supposed to basically be "-C prefer-dynamic"?

**Comment by alexcrichton** [comment]
Ok, if you want to use Rust projects with shared libraries amongst one another, you're basically exiting the world of the compiler and Cargo in terms of dependency resolution because they assume that "C libraries" (aka cdylibs) have their own build system and are accounted for separately.

You probably don't want `-C prefer-dynamic` as it's related to dylibs, not cdylibs. It sounds like you may be pretty far in the weeds though, so perhaps you may want to step back for a second? First off, shared libraries in Rust do not work well in terms of `extern crate foo` pulling in `foo` as a shared library. It works, but only barely. You probably want to statically link all your Rust dependencies as it won't really have that much of a footprint in practice.

The cdylib crate type is intended to produce a dynamic library to be loaded into an application, not be used as a dependency in Rust. That is, if you were to use one from Rust, you'd basically have to redefine everything with `extern { ... }` and have lots of unsafe glue between the two. This isn't really idiomatic Rust either so I wouldn't necessarily recommend it.

Basically if a rust crate depends on another, the crate depended on should *not* be a cdylib, otherwise you're probably going to run into pain as it's not really what they're intended for.

**Comment by MarkSwanson** [comment]
Ok, dropping cdylib for my current task. Will use it for another...

Note that I do statically link all other dependencies besides the common DSO and Rust std.

Wrt footprint... it's a really big deal in our environment. I'm hopeful that I'll be able to talk about why in more detail.

**Comment by alexcrichton** [comment]
Oh interesting! We *should* be able to do basically any reasonable linkage scenario, although not through Cargo but rather through the compiler. I can certainly try to help out if it's needed!

**Comment by prasannavl** [comment]
@alexcrichton - just been trying to find more info on the `cdylib`, but it makes me wonder - since it is assumed that using it leaves the Rust world for all practical purposes, and it's goal is to ease being used by C, can this be even taken further such that, what's exposed is similar in nature to how an rlib is exposed to another? 

By that I mean, add `#[no-mange]` along with `extern` automatically to all public functions. Perhaps, even a `repr` for public structs, where it is reasonable to do so. And separate module namespaces automatically with perhaps `_`, so most safe libraries (which do not rely on specific representations, or other unsafe features) can just be used from C by convention, with almost no changes? .. The ones that do use them are inherently unsafe - so it becomes the responsibility of the libraries to ensure it's correct anyway. 

While it may sound counter-intuitive at first, I wonder if there are any major disadvantages to do so. Making this kind of a friction-less escape hatch to use from C for many simple cases.

**Comment by alexcrichton** [comment]
@prasannavl what you're mentioning sounds like a larger change to the language rather than something that Cargo would do? That may be best discussed on the RFC repository rather than here perhaps?

**Comment by prasannavl** [comment]
@alexcrichton - You're right, of course. I've followed Rust for a long time now, but I'm fairly new to actually using it in real projects, and as such didn't feel my hands-on experience with the language was sufficient enough to start proposing RFCs.

So just wanted to throw it in some where of relevance to see if it made sense to someone living in the ecosystem, and if it's worth exploring further. And found this thread which seemed to be so :)
